apps/passwd.c

```diff
@@ -343,30 +331,48 @@ static char *md5crypt(const char *passwd, const char *magic, const char *salt)
 
     out_buf[0] = 0;
     magic_len = strlen(magic);
+    OPENSSL_strlcpy(ascii_magic, magic, sizeof ascii_magic);
+#ifdef CHARSET_EBCDIC
+    if ((magic[0] & 0x80) != 0)    /* High bit is 1 in EBCDIC alnums */
```

Comment:
It is an attempt to automatically detect input characters.  `magic` is a constant string that we have control over.  So far, it's "1", "apr1", "5", "6" or the empty string.  In EBCDIC, small letters are 0x81 to 0xA9 and digits are 0xF0 to 0xF9.  So yeah, sure, if someone invents a password format where the first character of the algorithm indicator isn't alphanumeric, then we'll get in trouble.  I don't think that's very likely, though.